botocore/http_session.py

```diff
@@ -0,0 +1,172 @@
+                 max_pool_connections=MAX_POOL_CONNECTIONS,
+    ):
+        self._verify = verify
+        self._proxies = proxies or {}
+        if timeout is None:
+            timeout=DEFAULT_TIMEOUT
+        if not isinstance(timeout, (int, float)):
+            timeout = Timeout(connect=timeout[0], read=timeout[1])
+        self._timeout = timeout
+        self._max_pool_connections = max_pool_connections
+        self._proxy_managers = {}
+        self._http_pool = PoolManager(
+            strict=True,
+            timeout=self._timeout,
+            maxsize=self._max_pool_connections,
+        )
+
+    def _get_proxy_manager(self, proxy_url):
+        proxy_url = fix_proxy_url(proxy_url)
+        if proxy_url not in self._proxy_managers:
+            proxy_headers = get_proxy_headers(proxy_url)
+            self._proxy_managers[proxy_url] = proxy_from_url(
+                proxy_url,
+                strict=True,
+                timeout=self._timeout,
+                proxy_headers=proxy_headers,
+                maxsize=self._max_pool_connections
+            )
+
+        return self._proxy_managers[proxy_url]
+
+    def _path_url(self, parsed_url):
+        path = parsed_url.path
+        if not path:
+            path = '/'
+        if parsed_url.query:
+            path = path + '?' + parsed_url.query
+        return path
+
+    def _verify_cert(self, conn, url, verify):
+        if url.lower().startswith('https') and verify:
+            conn.cert_reqs = 'CERT_REQUIRED'
+            conn.ca_certs = get_cert_path(verify)
+        else:
+            conn.cert_reqs = 'CERT_NONE'
+            conn.ca_certs = None
+
+    def send(self, request):
+        try:
+            parsed_url = urlparse(request.url)
+            proxy = self._proxies.get(parsed_url.scheme)
+            url = self._path_url(parsed_url)
+
+            if proxy:
+                if parsed_url.scheme == 'http':
+                    url = request.url
+                connection_manager = self._get_proxy_manager(proxy)
+            else:
+                connection_manager = self._http_pool
+
+            conn = connection_manager.connection_from_url(request.url)
+            self._verify_cert(conn, request.url, self._verify)
+            urllib_response = conn.urlopen(
+                method=request.method,
+                url=url,
+                body=request.body,
+                headers=request.headers,
+                retries=False,
+                assert_same_host=False,
+                preload_content=False,
+                decode_content=False,
+            )
+
+            http_response = AWSResponse(
+                url=request.url,
+                raw=urllib_response,
+                status_code=urllib_response.status,
+                headers=urllib_response.headers,
+            )
+
+            if not request.stream_output:
+                # Cause the raw stream to be exhausted immediatly
+                http_response.content
```

Comment:
What's the advantage of doing this vs. using the `preload_content` param?

Reply:
Not sure if there is an advantage. This is just how requests did it.

Reply:
Yes I'm aware requests does it like this.  I always found it odd because it's relying on attribute access to trigger external side effects, and if `preload_content` is identical it seems like that's a more explicit (and preferable) way of doing this.

Reply:
Looks like urllib3 will never preload the content if the response is chunked. Considering this we'll have to do it externally.